---
title: "WGA_SNV_hg38_07.25 - Whole Genome SNV analysis for hg38 builds"
author: "Layla Ahmed"
date: "2025-06-24"
output: html_document
---

```{r}
#You are currently using the Single Nucleotide Variant (SNV) filter code specifically for Whole Genome Analysis in hg38.
#If you want to specifically look for panel genes in your file, please go to PGA_SNV_hg38.
#If you are starting from the original TCAG file, please go to File -> Import Dataset -> Base (readr) as the file upload codeblock in this script will struggle to upload such a huge file.
#Please name your file 'patientdata' so the script can recognise it.

setup <- function(){
  messages <- c(
  "You are currently using the Single Nucleotide Variant (SNV) filter code specifically for Whole Genome Analysis.",
  "If you want to specifically look for panel genes in your file, please go to PGA_SNV_hg38.",
  "If you are starting from the original TCAG file, please go to File -> Import Dataset -> Base (readr) as the file upload codeblock in this script will struggle to upload such a huge file.",
  "Please name your file 'patientdata' so the script can recognise it."
)

# Loop through and print each with a pause
for (msg in messages) {
  cat(msg, "\n")
  Sys.sleep(2)  # Adjust seconds to your preference
}

  # Helper function for CRAN packages
install_and_load <- function(pkg) {
  if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
    install.packages(pkg)
    library(pkg, character.only = TRUE)
  } else {
    library(pkg, character.only = TRUE)
  }
}


install_and_load_bioc <- function(pkg) {
  if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
    if (!require("BiocManager", quietly = TRUE)) {
      install.packages("BiocManager")
      library(BiocManager)
    }
    BiocManager::install(pkg)
    library(pkg, character.only = TRUE)
  } else {
    library(pkg, character.only = TRUE)
  }
}   



# CRAN packages
cran_packages <- c(
  "data.table",
  "conflicted",
  "tidyverse",
  "writexl",
  "dplyr",
  "readr",
  "readxl",
  "openxlsx"
)

# Bioconductor packages
bioc_packages <- c(
  "AnnotationDbi",
  "org.Hs.eg.db"
)

# Install and load CRAN packages
lapply(cran_packages, install_and_load)

# Install and load Bioconductor packages
lapply(bioc_packages, install_and_load_bioc)

#For maths later on
count_removed_and_remaining_rows <<- function(original_data, modified_data) {
  removed_rows <- nrow(original_data) - nrow(modified_data)
  remaining_rows <- nrow(modified_data)
#Apparently dplyr and datatable have last as a function so i just changed it to lastlast (now everybody go chop breakfast)
lastlast <- function(x){x[[length(x)]]}
  
  cat(paste("This action removed", removed_rows, "rows.\n"))
  cat(paste("There are currently", remaining_rows, "rows remaining.\n"))
}

allDups = function(value){duplicated(value) | duplicated(value, fromLast = TRUE)} 

#Set up creation function so that runs automatically!

  Dnumber <- as.character(readline(prompt ="Please write the Sample Number:"))
   Dnumber_toremove_colon <<- paste0("^",Dnumber,"\\:")
  Dnumber_toremove_period <<- paste0("^",Dnumber,"\\.")
  invisible(assign("Dnumber", Dnumber, envir = .GlobalEnv))
  
FilterStep <<- list(
  Step = character(0),  # Empty character vector for Step names
  Count = numeric(0)    # Empty numeric vector for Count values
)
  

}

file_choice <- function(){
  if ("patientdata" %in% ls(envir = .GlobalEnv)) {
    print ("File named 'patientdata' found.")
    patientfile_cmd()
  } else {
    print ("'patientdata' is missing, please select a file:")
    patientfile <<- file.choose()
    patientfile_ext <<-  tools::file_ext(patientfile)
    if(patientfile_ext == "csv"){
      patientdata <<- read.csv(patientfile, stringsAsFactors = FALSE)
    } else if(patientfile_ext == "tsv"){
      patientdata <<- read.delim(patientfile, sep = "\t", strip.white = TRUE, stringsAsFactors = FALSE)
    } else if(patientfile_ext == "xlsx" | patientfile_ext == "xls"){
      # List out all the sheels in given Excel file
      sheetlist <<- excel_sheets(patientfile)
      # Print the sheet list and corresponding numbers
      cat("Choose the sheet you want to load in; each number corresponds with the adjacent sheet:\n")
      for (i in 1:length(sheetlist)) {
        cat(i, " - ", sheetlist[i], "\n")
      }
      input <- readline("Enter a number (or 'q' to quit): ")
      # Check if the input is 'q' (to quit)
      if (tolower(input) == "q") {
        break  # Exit the loop if 'q' is entered
      }
      # Convert the input to numeric
      num <- as.numeric(input)
      # Check if the input is a valid number and within the valid range
      if (!is.na(num) && num >= 1 && num <= length(sheetlist)) {
        KeepMe <<- as.character(sheetlist[num])
        KeepMeChara <- toString(KeepMe)
      } 
      else {
        cat("Invalid input. Please enter a valid number between 1 and ", length(sheetlist), ".\n")
      }
      
      # Print the matched values
      cat("This is the sheet we're keeping:", KeepMe, "\n")
      #then filter
      sheet_name <<- as.character(KeepMe)
      patientdata <<- read_xlsx(patientfile, sheet = sheet_name)
    }
    
    else {
      cat("Please pick a file in tabular format! Circling back...\n")
      file()
    }
    patientfile_cmd()
  }

}

#input the patient file
patientfile_cmd <- function(){
    dup_cols <<- names(patientdata)[duplicated(names(patientdata))]
    # Append a suffix to duplicate column names to make them unique
    for (col in dup_cols) {
      indices <- which(names(patientdata) == col)
      names(patientdata)[indices] <- paste(col, indices, sep = "_")
    }
    patientdata <<- patientdata %>%
      separate_rows(gene_symbol, sep = ",")
      #Make sure the columns dont have the sample name inside
    names(patientdata) <<- sub(Dnumber_toremove_colon, "", names(patientdata))
    names(patientdata) <<- sub(Dnumber_toremove_period, "", names(patientdata))
    currentdata <<- patientdata
    if ("spliceAI_DS_AL" %in% colnames(currentdata)){
    currentdata$sAI_delta <<- currentdata$spliceAI_DS_AG + currentdata$spliceAI_DS_AL + currentdata$spliceAI_DS_DG + currentdata$spliceAI_DS_DL
    }
    currentdata <<- currentdata
    InitialRow <<- nrow(currentdata)
    InitialText <<- "Initial Number of Variants"
    FilterStep$Step <<- c(InitialText)
    FilterStep$Count <<- c(InitialRow)
    rm(patientdata, envir = .GlobalEnv )
   
}

highqualityfilter  <- function(){
  cat("Quality filtering will now commence!\n")
  #Filter FILTER = PASS; all variants that pass the filter check
  cat("Filtering for variants that PASS the filter check.\n")
  PASS_filtered_data <- currentdata[currentdata$FILTER%in%'PASS',]
  count_removed_and_remaining_rows(currentdata, PASS_filtered_data)
  currentdata <- PASS_filtered_data
  currentdata$newDP <- as.numeric(currentdata$AD_ALT)+as.numeric(currentdata$AD_REF)
  highquality_filtered_data <- currentdata[as.numeric(currentdata$newDP)>=7,]
  highquality_filtered_data <- highquality_filtered_data[as.numeric(highquality_filtered_data$newDP)<=120,]
  Quality_Row <<- nrow(highquality_filtered_data)
  Quality_Text <<- "Quality Filter: FILTER = PASS & 7 > Depth > 120"
  currentdata <<- highquality_filtered_data
   FilterStep$Step <<- c(FilterStep$Step, Quality_Text)
  FilterStep$Count <<- c(FilterStep$Count, Quality_Row)
  #wb <- openxlsx::createWorkbook()
  #sheet_name <<- as.character("Quality_Filtered_Variants")
 # addWorksheet(wb, sheet_name)
 # final_file <- paste0(Dnumber, "_qualityfilter.xlsx")
 # writeData(wb, sheet = sheet_name, x = currentdata)
  # Save the Excel file
 # saveWorkbook(wb, file = final_file)

}


## 0.5% Filtering of Population #

populationfiltering <- function(){
  cat("Population filtering will now commence!\n")
  dup_cols <- which(duplicated(names(currentdata)))
  if (length(dup_cols) > 0) {
    names(currentdata)[dup_cols] <- paste(names(currentdata)[dup_cols], seq_along(dup_cols), sep = "_")
  }
  if ("1000g_all" %in% colnames(currentdata)){
    currentdata$"1000g_all" <- as.numeric(currentdata$"1000g_all")
  popfreq_r_05_100_data <<- currentdata[ currentdata$"1000g_all" <= 0.005,] 
  } else if (("A1000g_freq_max" %in% colnames(currentdata)) && ("freq_max" %in% colnames(currentdata))){
  currentdata$A1000g_freq_max <- as.numeric(currentdata$A1000g_freq_max)
  currentdata$freq_max <- as.numeric(currentdata$freq_max)
  popfreq_r_05_100_data <<- currentdata[currentdata$A1000g_freq_max <= 0.005 & currentdata$freq_max <= 0.005,] 
  } else if (("A1000g_freq_max" %in% colnames(currentdata))){
  currentdata$A1000g_freq_max <- as.numeric(currentdata$A1000g_freq_max)
  popfreq_r_05_100_data <<- currentdata[currentdata$A1000g_freq_max <= 0.005,] 
  } else if (("freq_max" %in% colnames(currentdata))){
  currentdata$freq_max <- as.numeric(currentdata$freq_max)
  popfreq_r_05_100_data <<- currentdata[currentdata$freq_max <= 0.005,] 
  }
  count_removed_and_remaining_rows(currentdata, popfreq_r_05_100_data)
  RareRow <<- nrow(popfreq_r_05_100_data)
  RareText <<- " Frequency Filtering (1000 Genomes & gnomAD <= 0.05%)"
  FilterStep$Step <<- c(FilterStep$Step, RareText)
  FilterStep$Count <<- c(FilterStep$Count, RareRow)
   wb <- openxlsx::createWorkbook()
  sheet_name <<- as.character("popfreq_r_05_100")
  addWorksheet(wb, sheet_name)
  final_file <- paste0(Dnumber, "_popfreq_05_100.xlsx")
  writeData(wb, sheet = sheet_name, x = popfreq_r_05_100_data)
  
  if (file.exists(final_file)) {
    counter <<- 1
    dup_file <<- paste0(Dnumber, "_", counter, "_popfreq_05_100.xlsx")

    while (file.exists(dup_file)) {
        counter <<- counter + 1
        dup_file <<- paste0(Dnumber,"_", counter, "_popfreq_05_100.xlsx")
    }
    saveWorkbook(wb, file = dup_file)
    cat(paste0("File saved as '", dup_file, "'.\n"))
} else {
    saveWorkbook(wb, file = final_file)
    paste0(final_file, "saved successfully!\n")
  
}
 
  currentdata <<- popfreq_r_05_100_data
}

### Type Filtering

varianttype_filtring<- function(){
  cat("Variant type filtering will now commence!\n")
  RemoveforGenic <-  c('intergenic','ncRNA_exonic','ncRNA_intronic','ncRNA_splicing','ncRNA_UTR3','ncRNA_UTR5','ncRNA_exonic;ncRNA_splicing', 'downstream','upstream','upstream;downstream','UTR3','UTR5','UTR5;UTR3')
  genic_variants <<-currentdata[!currentdata$typeseq_priority%in%RemoveforGenic,]
  GenicRow <<- nrow(genic_variants)
  GenicText <<- "Genic Variants"
  FilterStep$Step <<- c(FilterStep$Step, GenicText)
  FilterStep$Count <<- c(FilterStep$Count, GenicRow)
  currentdata <<- genic_variants
#  wb <- openxlsx::createWorkbook()
 # sheet_name <<- as.character("Genic_Variants")
 # addWorksheet(wb, sheet_name)
 # final_file <- paste0(Dnumber, "_genic_variants.xlsx")
 # writeData(wb, sheet = sheet_name, x =  genic_variants)
  # Save the Excel file
 # saveWorkbook(wb, file = final_file)
  
  }

#Filering for variants that have a Splice AI scoreabove cutoff#
spliceAI_filter <- function(){
  
  dup_cols <- which(duplicated(names(currentdata)))
  if (length(dup_cols) > 0) {
    names(currentdata)[dup_cols] <<- paste(names(currentdata)[dup_cols], seq_along(dup_cols), sep = "_")
  }
  #then coerce the required columns into a numeric format
  cat("Splice AI filtering - cutoff is 0.1.\n")
  
  spliceAI_data <<- currentdata[currentdata$sAI_delta>= 0.1,]
  Splice_Text <- "Splice AI Filtering"
  Splice_Row <- nrow(spliceAI_data)
   FilterStep$Step <<- c(FilterStep$Step, Splice_Text)
  FilterStep$Count <<- c(FilterStep$Count, Splice_Row)
  } 


effect_filter <- function(){
  cat("Loss-of-Function only filtering will now commence!\n")
  dup_cols <- which(duplicated(names(currentdata)))
  if (length(dup_cols) > 0) {
    names(currentdata)[dup_cols] <<- paste(names(currentdata)[dup_cols], seq_along(dup_cols), sep = "_")}
  LOFkeep <- c('frameshift deletion', 'frameshift insertion', 'stoploss','stopgain')
  ClinVarkeep <- c('Pathogenic/Likely_pathogenic', 'Likely_pathogenic', 'Pathogenic')
  ClinVar_data <<- currentdata[currentdata$Clinvar_SIG%in%ClinVarkeep,]
  LOFonly_data <<- currentdata[currentdata$effect_priority%in%LOFkeep,]
  effect_data <<- rbind(ClinVar_data,LOFonly_data)
  effect_data <<- effect_data[!duplicated(effect_data), ]
  LOF_Row <<- nrow(effect_data)
  LOF_Text <<- "Effect Filtering (Variant Effect + ClinVar Significance"
  FilterStep$Step <<- c(FilterStep$Step, LOF_Text)
  FilterStep$Count <<- c(FilterStep$Count, LOF_Row)
}

#Predicitve Pathogenic Filtering#
pathogenicfiltering <- function(){
  allDups <<- function(value){duplicated(value) | duplicated(value, fromLast = TRUE)} 
  cat("Filtering via pathogenic filters will now commence with pre-set cut offs!\n")
    user_cutoffs <- list(
    sift_scorecutoff = 0.05,
    polyphen_scorecutoff = 0.8,
    ma_scorecutoff = 2,
    spx_dpsicutoff = 2,
    phylopPMam_avgcutoff = 1.5,
    phylopVert100_avgcutoff = 4,
    PROVEAN_scorecutoff = -2.5,
    CADD_phredcutoff = 15,
    REVEL_cutoff = 0.5
    )
    invisible(assign("user_cutoffs", user_cutoffs, envir = .GlobalEnv))
  
  
  dup_cols <- which(duplicated(names(currentdata)))
  if (length(dup_cols) > 0) {
    names(currentdata)[dup_cols] <- paste(names(currentdata)[dup_cols], seq_along(dup_cols), sep = "_")}
  predVote = suppressWarnings(cbind(as.numeric(currentdata$sift_score) <=  user_cutoffs$sift_scorecutoff, 
                                    as.numeric(currentdata$polyphen_score) >= user_cutoffs$polyphen_scorecutoff, 
                                    as.numeric(currentdata$ma_score) >= user_cutoffs$ma_scorecutoff,
                                    abs(as.numeric(currentdata$spx_dpsi)) <= user_cutoffs$spx_dpsicutoff,
                                    as.numeric(currentdata$phylopMam_avg) >= user_cutoffs$phylopPMam_avgcutoff,
                                    as.numeric(currentdata$phylopVert100_avg) >= user_cutoffs$phylopVert100_avgcutoff,
                                    as.numeric(currentdata$PROVEAN_score) <= user_cutoffs$PROVEAN_scorecutoff,
                                    as.numeric(currentdata$CADD_phred) >= user_cutoffs$CADD_phredcutoff,
                                    as.numeric(currentdata$REVEL_score) >= user_cutoffs$REVEL_cutoff))
  #ind = rowSums(predVote, na.rm=TRUE)
  currentdata$ind <<- rowSums(predVote, na.rm = TRUE)
  pathogenicfilter_data <<- dplyr::filter(currentdata, ind >= 2)
  pathogenicfilter_data <<- pathogenicfilter_data %>%
  dplyr::filter(!(effect_priority %in% c('nonframeshift insertion', 'nonframeshift deletion') & str_count(dbsnp_region, ",") >= 4)) 
  if ("Clinvar_SIG" %in% colnames(currentdata) && "Clinvar_ReviewStatus" %in% colnames(currentdata) ) {
  pathogenicfilter_data <<- pathogenicfilter_data %>%
  dplyr::filter(!(Clinvar_SIG %in%c('Benign') & Clinvar_ReviewStatus %in% c('no_assertion_criteria_provided')))
  }
  pathogenicfilter_data$ind <<- NULL
  pred.path_Row <<- nrow(pathogenicfilter_data)
  pred.path_Text <<- "Passed Predictor Filtering"
  FilterStep$Step <<- c(FilterStep$Step, pred.path_Text)
  FilterStep$Count <<- c(FilterStep$Count, pred.path_Row)
  strong_variants <<- rbind(effect_data,pathogenicfilter_data,spliceAI_data)
  Strong_Var_Row <<- nrow(strong_variants)
  Strong_Var_Text <<- "Predicted as damaging, deleterious, conserved or splicing"
  FilterStep$Step <- c(FilterStep$Step, Strong_Var_Text)
  FilterStep$Count <- c(FilterStep$Count, Strong_Var_Row)
  strong_variants <<- strong_variants[!duplicated(strong_variants), ]
  strong_variants <<- strong_variants[order(strong_variants$gene_symbol),] 
}

scoringstrongSNVs <- function(){
   strong_variants$VariantScore <<- 0
   # Remove the row if it is exactly "Benign"
# Sets up cutoffs that earn 1 mark
user_cutoffs <- list(
  sift_scorecutoff = 0.05,
  polyphen_scorecutoff = 0.8,
  ma_scorecutoff = 2,
  spx_dpsicutoff = 2,
  phylopPMam_avgcutoff = 1.5,
  phylopVert100_avgcutoff = 4,
  PROVEAN_scorecutoff = -2.5,
  CADD_phredcutoff = 15,
  REVEL_cutoff = 0.5
)
# Check and rename duplicate column names
dup_cols <<- which(duplicated(names(strong_variants)))
if (length(dup_cols) > 0) {
  names(strong_variants)[dup_cols] <- paste(names(strong_variants)[dup_cols], seq_along(dup_cols), sep = "_")
}

# Calculate full cutoff scores
predVote = suppressWarnings(cbind(
  as.numeric(strong_variants$sift_score) <= user_cutoffs$sift_scorecutoff, 
  as.numeric(strong_variants$polyphen_score) >= user_cutoffs$polyphen_scorecutoff, 
  as.numeric(strong_variants$ma_score) >= user_cutoffs$ma_scorecutoff,
  abs(as.numeric(strong_variants$spx_dpsi)) <= user_cutoffs$spx_dpsicutoff,
  as.numeric(strong_variants$phylopMam_avg) >= user_cutoffs$phylopPMam_avgcutoff,
  as.numeric(strong_variants$phylopVert100_avg) >= user_cutoffs$phylopVert100_avgcutoff,
  as.numeric(strong_variants$PROVEAN_score) <= user_cutoffs$PROVEAN_scorecutoff,
  as.numeric(strong_variants$CADD_phred) >= user_cutoffs$CADD_phredcutoff,
  as.numeric(strong_variants$REVEL_score) >= user_cutoffs$REVEL_cutoff
))

# Assign full scores
strong_variants$ind <<- rowSums(predVote, na.rm = TRUE)

strong_variants$VariantScore <<- strong_variants$ind  

#Loop that looks at some other significant scores
if ("Clinvar_SIG" %in% colnames(currentdata)) {
 for (i in 1:nrow(strong_variants)) {
  if (grepl("pathogenic", strong_variants$Clinvar_SIG[i], ignore.case = TRUE)) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 30
  } else if (grepl("Conflicting", strong_variants$Clinvar_SIG[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 2
  } else if (grepl("Uncertain", strong_variants$Clinvar_SIG[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 2
  } 
 }
}
if ("effect_priority" %in% colnames(currentdata)) {
 for (i in 1:nrow(strong_variants)) {
  if (grepl("frameshift", strong_variants$effect_priority[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 15
  }
   if (grepl("stopgain", strong_variants$effect_priority[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 15
  }
   if (grepl("startloss", strong_variants$effect_priority[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 5
   }
   if (grepl("stoploss", strong_variants$effect_priority[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 5
   }
   if (grepl("inframe", strong_variants$effect_priority[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 5
   }
 }
}

if ("sAI_delta" %in% colnames(currentdata)) {
 for (i in 1:nrow(strong_variants)) {
 if (any(strong_variants$sAI_delta >= 0.2, na.rm = TRUE)) {
  strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 1
} else if (any(strong_variants$sAI_delta >= 0.1, na.rm = TRUE)) {
  strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 0.5
}
 }
}
  # 6. MPO column
  if (grepl("vision", strong_variants$MPO[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 1
  }

  # 7. HPO column
  if (grepl("eye", strong_variants$HPO[i])) {
    strong_variants$VariantScore[i] <- strong_variants$VariantScore[i] + 1
  }
  

strong_variants <<- strong_variants[order(-strong_variants$VariantScore), ]
   
  
FilterRows <- data.frame(
  Step = FilterStep$Step,
  Count = FilterStep$Count,
  stringsAsFactors = FALSE
)  
  
  
wb <- openxlsx::createWorkbook()
  sheet_name_all <<- as.character("strong_variants")
  sheet_name_snv <<- as.character("SNV_Filter_History")
  addWorksheet(wb, sheet_name_all)
  addWorksheet(wb, sheet_name_snv)
  final_file <- paste0(Dnumber, "_Strong_variants.xlsx")
  writeData(wb, sheet = sheet_name_all, x = strong_variants)
  writeData(wb, sheet = sheet_name_snv, x = FilterRows)
  # Save the Excel file
if (file.exists(final_file)) {
    counter <<- 1
    dup_file <<- paste0(Dnumber, "_", counter, "_Strong_variants.xlsx")

    while (file.exists(dup_file)) {
        counter <<- counter + 1
        dup_file <<- paste0(Dnumber,"_", counter, "_Strong_variants.xlsx")
    }
    saveWorkbook(wb, file = dup_file)
    cat(paste0("File saved as '", dup_file, "'.\n"))
} else {
    saveWorkbook(wb, file = final_file)
    paste0(final_file, "saved successfully!\n")
  
}
  }
  
setup()
file_choice()
gc()
highqualityfilter()
gc()
populationfiltering()
gc()
varianttype_filtring()
gc()
spliceAI_filter()
gc()
effect_filter()
gc()
pathogenicfiltering()
gc()
scoringstrongSNVs()
gc()
  
```

